File: src/main/java/aoc2019/Day08.java
package aoc2019;

import java.util.Arrays;
import java.util.Comparator;
import java.util.function.Function;
import java.util.stream.IntStream;

/*
 * Day 8: Space Image Format
 * https://adventofcode.com/2019/day/8
 */
public class Day08 {

	private static final int COLS = 25;
	private static final int ROWS = 6;
    private static final int LAYER_SIZE = COLS*ROWS;

    private static final int BLACK = 0;
    private static final int WHITE = 1;
    private static final int TRANSPARENT = 2;
    
	public static void main(String[] args) throws Exception {
        new Day08().solve();
    }

	private int[] input;

    void solve() throws Exception {
    	System.out.println("=== input ===");
    	
        input = Util.lines("input08.txt").get(0).chars().map(n -> n-'0').toArray();
        int layers = input.length/LAYER_SIZE;
        System.out.println(String.format("found %d layers", layers));
        
        System.out.println("=== part1 ===");
        
        Function<Integer, Long> zeroes = n -> layer(n).filter(i -> i==0).count();        
        int minZeroes = IntStream.range(0, layers).boxed().min(Comparator.comparing(zeroes)).get();  
        System.out.println("min zeroes on layer " + minZeroes);
        
        long ones = layer(minZeroes).filter(i -> i==1).count();
        long twos = layer(minZeroes).filter(i -> i==2).count();
        var result = ones*twos;
        System.out.println(result);
        
        System.out.println("=== part2 ===");

//        int i = layers-1;
//        var img = layer(i).toArray();
//        for (; i>=1; --i) merge(img, layer(i-1).toArray());
        
        // all layer arrays in reversed order
        var allLayers = IntStream.range(0, layers).mapToObj(n -> layer(layers-n-1).toArray());
        var img = allLayers.reduce(new int[LAYER_SIZE], (lower, upper) -> merge(lower, upper));
        
        printImage(img);
    }

	void printImage(int[] img) {
		for (int r=0; r<ROWS; ++r) {
        	for (int c=0; c<COLS; ++c) System.out.print(img[r*COLS+c] == WHITE? '#':' ');
        	System.out.println();
        }
	}
   
	IntStream layer(int n) {
		return Arrays.stream(input).skip(n*LAYER_SIZE).limit(LAYER_SIZE);
	}
	
	int overlay(int lower, int upper) {
		return (upper == TRANSPARENT)? lower: upper;
	}
	
	int[] merge(int[] lower, int[] upper) {
		for (int i=0; i<upper.length; ++i) lower[i] = overlay(lower[i], upper[i]); 
		return lower;
	}
}

File: src/main/go/aoc2019/day08.go
package main

import (
	"fmt"
	"time"
)

// Run08 Day 8: Space Image Format
func Run08() {
	t := time.Now()
	fmt.Println("--- part 1")
	day08Part1()

	fmt.Println("--- part 2")
	day08Part2()

	fmt.Printf("=== end (%d ms) ===\n", time.Now().Sub(t).Milliseconds())
}

func day08Part1() {
	s0 := ReadInputStrings("input08.txt")[0]
	img := make([]int, len(s0))
	for i, si := range s0 {
		img[i] = int(si) - '0'
	}
	fmt.Println(img[0:10])
	layersize := 25 * 6
	nlayers := len(img) / layersize
	fmt.Println("layers in image: ", nlayers)

	layer := func(l int) []int {
		return img[l*layersize : (l+1)*layersize]
	}

	noOfDigits := func(layer []int, digit int) (noz int) {
		for _, d := range layer {
			if d == digit {
				noz++
			}
		}
		return
	}

	r, z := 0, layersize
	for l := 0; l < nlayers; l++ {
		x := noOfDigits(layer(l), 0)
		if x < z {
			r = l
			z = x
		}
	}
	fmt.Println("result: ", noOfDigits(layer(r), 1)*noOfDigits(layer(r), 2))
}

func day08Part2() {
	s0 := ReadInputStrings("input08.txt")[0]
	img := make([]int, len(s0))
	for i, si := range s0 {
		img[i] = int(si) - '0'
	}
	layersize := 25 * 6
	nlayers := len(img) / layersize
	layer := func(l int) []int {
		return img[l*layersize : (l+1)*layersize]
	}

	last := layer(nlayers - 1)
	for l := nlayers - 2; l >= 0; l-- {
		for i, p := range layer(l) {
			if p != 2 {
				last[i] = p
			}
		}
	}

	for y := 0; y < 6; y++ {
		for x := 0; x < 25; x++ {
			fmt.Print(string(" #"[last[y*25+x]]))
		}
		fmt.Println()
	}
}
